File: gameSrc/localization.js
"use strict";

var debug = require("./debug");

/**
 * Holds all available languages.
 */
var languages = {};

/**
 * The current active language.
 */
var selected = null;

/**
 * Registers a language object. The properties of the object will be the keys and its
 * values the localized string for the key.
 */
exports.registerLanguage = function (key, obj) {
  if (!key || !obj) debug.logCritical("IllegalArgumentException");
  if (languages.hasOwnProperty(key)) debug.logCritical("LanguageAlreadyRegisteredException");

  // copy keys and values to a fresh object
  var newLang = {};
  var wordKey;
  for (wordKey in languages) {
    if (languages.hasOwnProperty(wordKey)) newLang[wordKey] = obj[wordKey];
  }

  // register it
  languages[key] = newLang;
};

/**
 * Selects a language by it's key.
 */
exports.selectLanguage = function (key) {
  if (!languages.hasOwnProperty(key)) debug.logCritical("UnknownLanguageException");
  selected = languages[key];
};

/**
 * Returns the localized string of a given identifier.
 */
exports.forKey = function (key) {
  if (!selected) return key;
  var str = selected[key];
  return (str || key);
};

File: gameSrc/net/wolfTec/utility/Localization.java
package net.wolfTec.utility;

import net.wolfTec.Constants;
import org.stjs.javascript.JSCollections;
import org.stjs.javascript.JSObjectAdapter;
import org.stjs.javascript.Map;

import java.util.Iterator;

public class Localization {

    public static final String LOG_HEADER = Constants.logHeader("localization");

    /**
     * Holds all available languages.
     */
    private Map<String, Map<String, String>> languages = JSCollections.$map();

    /**
     * The current active language.
     */
    private Map<String, String> selected;

    /**
     * Registers a language object. The properties of the object will be the keys and its
     * values the localized string for the key.
     */
    public void registerLanguage(String key, Map<String, String> obj) {
        if (key == null || obj == null) Debug.logCritical(LOG_HEADER, "IllegalArgumentException");
        if (JSObjectAdapter.hasOwnProperty(languages, key)) {
            Debug.logCritical(LOG_HEADER, "LanguageAlreadyRegisteredException");
        }

        // copy keys and values to a fresh object
        Map<String, String> newLang = JSCollections.$map();
        Iterator<String> langIt = languages.iterator();
        while (langIt.hasNext()) {
            String langItemKey = langIt.next();
            newLang.$put(langItemKey, obj.$get(langItemKey));
        }

        // register it
        languages.$put(key, newLang);
    }

    /**
     * Selects a language by it's key.
     */
    public void selectLanguage(String key) {
        if (!JSObjectAdapter.hasOwnProperty(languages, key)) throw new IllegalArgumentException("unknown language");
        selected = languages.$get(key);
    }

    /**
     * Returns the localized string of a given identifier.
     */
    public String forKey(String key) {
        if (selected == null) return key;
        String str = selected.$get(key);
        return str != null ? str : key;
    }
}
